solang-parser/src/helpers/ord.rs

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements `PartialOrd` and `Ord` for some parse tree data types, following the
+//! [Solidity style guide][ref].
+//!
+//! [ref]: https://docs.soliditylang.org/en/latest/style-guide.html
+
+use crate::pt;
+use std::cmp::Ordering;
+
+macro_rules! impl_with_cast {
+    ($($t:ty),+) => {
+        $(
+            impl $t {
+                #[inline]
+                const fn as_discriminant<'a>(&'a self) -> &'a u8 {
+                    // SAFETY: See <https://doc.rust-lang.org/stable/std/mem/fn.discriminant.html#accessing-the-numeric-value-of-the-discriminant>
+                    // and <https://doc.rust-lang.org/reference/items/enumerations.html#pointer-casting>
+                    //
+                    // `$t` must be `repr(u8)` for this to be safe.
+                    unsafe { &*(self as *const Self as *const u8) }
+                }
+            }
+
+            impl PartialOrd for $t {
+                #[inline]
+                fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+                    PartialOrd::partial_cmp(self.as_discriminant(), other.as_discriminant())
+                }
+            }
+
+            impl Ord for $t {
+                #[inline]
+                fn cmp(&self, other: &Self) -> Ordering {
+                    Ord::cmp(self.as_discriminant(), other.as_discriminant())
+                }
+            }
+        )+
+    };
+}
+
+// SAFETY: every type must be `repr(u8)` for this to be safe, see comments in macro implementation.
+impl_with_cast!(pt::Visibility, pt::VariableAttribute, pt::FunctionAttribute);
```

Comment:
I do totally get why you've done it this way, but I'd like to avoid `unsafe {}`. What do you think of the following?

```suggestion
impl pt::Visibility {
    fn solidity_order(&self) -> u8 {
        match self {
            pt::Visibility::External(..) => 0,
            pt::Visibility::Internal(..) => 1,
            pt::Visibility::Public(..) => 2,
            pt::Visibility::Private(..) => 3,
        }
    }
}

impl PartialOrd for pt::Visibility {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.solidity_order().cmp(&other.solidity_order()))
    }
}

impl Ord for pt::Visibility {
    fn cmp(&self, other: &Self) -> Ordering {
        self.solidity_order().cmp(&other.solidity_order())
    }
}

impl pt::VariableAttribute {
    fn solidity_order(&self) -> u8 {
        match self {
            pt::VariableAttribute::Visibility(..) => 0,
            pt::VariableAttribute::Constant(..) => 1,
            pt::VariableAttribute::Immutable(..) => 2,
            pt::VariableAttribute::Override(..) => 3,
        }
    }
}

impl PartialOrd for pt::VariableAttribute {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.solidity_order().cmp(&other.solidity_order()))
    }
}

impl Ord for pt::VariableAttribute {
    fn cmp(&self, other: &Self) -> Ordering {
        self.solidity_order().cmp(&other.solidity_order())
    }
}

impl pt::FunctionAttribute {
    fn solidity_order(&self) -> u8 {
        match self {
            pt::FunctionAttribute::Visibility(..) => 0,
            pt::FunctionAttribute::Mutability(..) => 1,
            pt::FunctionAttribute::Virtual(..) => 2,
            pt::FunctionAttribute::Immutable(..) => 3,
            pt::FunctionAttribute::Override(..) => 4,
            pt::FunctionAttribute::BaseOrModifier(..) => 5,
            pt::FunctionAttribute::Error(..) => 6,
        }
    }
}

impl PartialOrd for pt::FunctionAttribute {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.solidity_order().cmp(&other.solidity_order()))
    }
}

impl Ord for pt::FunctionAttribute {
    fn cmp(&self, other: &Self) -> Ordering {
        self.solidity_order().cmp(&other.solidity_order())
    }
}
```

Reply:
That's how I originally implemented it, though I didn't feel like it was really idiomatic. Seeing as the standard Discriminant doesn't implement Ord, I then tried this method which is recommended in the Rust docs. I believe it's also easier to maintain and has no overhead due to it being just a pointer cast.
But if you would like to avoid unsafe at all costs then I can change this, up to you.

Reply:
It's a matter of taste - let's go with your version. Thank you for the discussion.